File: 1-two-sum/Solution.cc
class Solution {
public:
    // O(n) Solution
    vector<int> twoSum(vector<int>& n, int target) {
        // (value, index) mapping
        unordered_map<int, int> M;
        for(int i=0; i<n.size(); i++){
            M[n[i]] = i;
        }
        
        vector<int> r;
        for(int i=0; i<n.size(); i++){
            auto it = M.find(target - n[i]);
            if(it != M.end() && it->second != i){
                r.push_back(i);                
                r.push_back(it->second);
                break;
            }
        }  
        return r;
    }
};

File: 1-two-sum/Solution.java
// O(n) Solution
public class Solution {
    public int[] twoSum(int[] nums, int target) {
        // (value, index) mapping
        HashMap<Integer, Integer> hmap = new HashMap<Integer, Integer>();
        for(int i=0; i<nums.length; i++){
            hmap.put(nums[i], i);
        }
        int[] r = new int[2];
        for(int i=0; i<nums.length; i++){
            int diff = target - nums[i];
            if(hmap.containsKey(diff)){
                int index = hmap.get(diff);
                if(index != i){
                    r[0] = i;
                    r[1] = index;
                    break;
                }
            }
        }
        return r;
    }
}